pkg/sfu/receiver.go

```diff
@@ -870,35 +867,38 @@ func (w *WebRTCReceiver) GetPrimaryReceiverForRed() TrackReceiver {
 		return w
 	}
 
-	if w.primaryReceiver.Load() == nil {
+	if w.redTransformer.Load() == nil {
 		pr := NewRedPrimaryReceiver(w, DownTrackSpreaderParams{
 			Threshold: w.lbThreshold,
 			Logger:    w.logger,
 		})
-		if w.primaryReceiver.CompareAndSwap(nil, pr) {
-			w.redPktWriter.Store(redPktWriteFunc(pr.ForwardRTP))
-			w.redSenderReportWriter.Store(redSenderReportWriteFunc(pr.ForwardRTCPSenderReport))
-		}
+		w.redTransformer.CompareAndSwap(nil, pr)
+	}
+	rt := w.redTransformer.Load()
+	if pr, ok := rt.(*RedPrimaryReceiver); ok {
+		return pr
 	}
-	return w.primaryReceiver.Load()
+	return nil
 }
 
 func (w *WebRTCReceiver) GetRedReceiver() TrackReceiver {
 	if w.isRED || w.closed.Load() {
 		return w
 	}
 
-	if w.redReceiver.Load() == nil {
+	if w.redTransformer.Load() == nil {
 		pr := NewRedReceiver(w, DownTrackSpreaderParams{
 			Threshold: w.lbThreshold,
 			Logger:    w.logger,
 		})
-		if w.redReceiver.CompareAndSwap(nil, pr) {
-			w.redPktWriter.Store(redPktWriteFunc(pr.ForwardRTP))
-			w.redSenderReportWriter.Store(redSenderReportWriteFunc(pr.ForwardRTCPSenderReport))
-		}
+		w.redTransformer.CompareAndSwap(nil, pr)
 	}
-	return w.redReceiver.Load()
+
+	rt := w.redTransformer.Load()
```

Comment:
this could be moved above the condition to avoid loading twice?

Reply:
Yeah, will do. Will just use the lock to synchronise. I think (I have to check) it could get called multiple times. Anyhow, will just use the struct lock to synchronise and simplify.

Reply:
Changed in this [commit](https://github.com/livekit/livekit/pull/3481/commits/14082bfa5151b2d4b8d8eedeba84e170f27fdf77).